src/catalog/catalog.cpp

```diff
@@ -0,0 +1,162 @@
+#include <memory>
+#include <unordered_map>
+#include <vector>
+
+#include "catalog/catalog.h"
+#include "catalog/database_handle.h"
+#include "loggers/catalog_logger.h"
+#include "storage/storage_defs.h"
+#include "transaction/transaction_manager.h"
+
+namespace terrier::catalog {
+
+std::shared_ptr<Catalog> terrier_catalog;
+
+Catalog::Catalog(transaction::TransactionManager *txn_manager) : txn_manager_(txn_manager) {
+  CATALOG_LOG_TRACE("Creating catalog ...");
+  Bootstrap();
+}
+
+DatabaseHandle Catalog::GetDatabaseHandle(db_oid_t db_oid) { return DatabaseHandle(this, db_oid, pg_database_); }
+
+std::shared_ptr<storage::SqlTable> Catalog::GetDatabaseCatalog(db_oid_t db_oid, table_oid_t table_oid) {
+  return map_[db_oid][table_oid];
+}
+
+void Catalog::Bootstrap() {
+  CATALOG_LOG_TRACE("Bootstrapping global catalogs ...");
+  // refer to OID assignment scheme in catalog.h
+  table_oid_t table_oid(DEFAULT_TABLE_OID);
+  col_oid_t col_oid(DEFAULT_COL_OID);
+  transaction::TransactionContext *txn = txn_manager_->BeginTransaction();
+  CATALOG_LOG_TRACE("Creating pg_database table ...");
+  CreatePGDatabase(txn, table_oid++, &col_oid);
+  CATALOG_LOG_TRACE("Creating pg_tablespace table ...");
+  CreatePGTablespace(txn, table_oid++, &col_oid);
+
+  BootstrapDatabase(txn, DEFAULT_DATABASE_OID);
+  txn_manager_->Commit(txn, BootstrapCallback, nullptr);
+  delete txn;
+}
+
+void Catalog::BootstrapDatabase(transaction::TransactionContext *txn, db_oid_t db_oid) {
+  // refer to OID assignment scheme in catalog.h
+  table_oid_t table_oid(DATABASE_CATALOG_TABLE_START_OID);
+  col_oid_t col_oid(DATABASE_CATALOG_COL_START_OID);
+  namespace_oid_t namespace_oid(PG_CATALOG_OID);
+
+  // create pg_namespace
+  table_oid_t pg_namespace_oid(table_oid++);
+  std::vector<Schema::Column> cols;
+  cols.emplace_back("oid", type::TypeId::INTEGER, false, col_oid++);
```

Comment:
I did not see how col_oid value is updated (preserved) after these increments.
I also think that (*start_col_oid)++ is not safe, e.g when we later do things in parallel, and try to create 2 user tables at the same time.

One solution to both of these, would be to have methods accessed via the database handle, that give you an oid value, e.g. something like GetNextColOid(). The increment needs to be atomic (so it could be lockless atomic or use a lock). You'd need functions to hand out oids for each range.